backend/game/views/player.py

```diff
@@ -0,0 +1,35 @@
+from rest_framework import viewsets, status, mixins
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+from django.shortcuts import get_object_or_404
+from game.models import Game, Player
+from game.serializers import PlayerSerializer
+from users.models import User
+
+
+class PlayerViewSet(mixins.CreateModelMixin,
+                    mixins.RetrieveModelMixin,
+                    mixins.DestroyModelMixin,
+                    mixins.ListModelMixin,
+                    viewsets.GenericViewSet):
+    queryset = Player.objects.all()
+    serializer_class = PlayerSerializer
+    permission_classes = [IsAuthenticated]
+
+    def create(self, request, *args, **kwargs):
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        # Retrieve game and user, return a 404 response if at least one doesn't exist
+        game_id, user_id = serializer.data.get('game'), serializer.data.get('user')
```

Comment:
Inutile de récupérer le user_id du serializer si on va de toute façon utiliser `request.user.id`

Reply:
Ici on utilise bien le user_id, et on vérifie si `user_id ==request.user.id`

Reply:
Ce que je voulais dire c'est qu'il est inutile de prendre en compte un éventuel champ `user_id` dans le corps de la requête, donc cette vérification n'a pas lieu d'être

Reply:
Pour moi la fonction devrait ressembler à ça, et dans ce cas il n'est pas possible de fournir un mauvais utilisateur dans le corps de la requête :

`    def create(self, request, *args, **kwargs):
        serializer = self.get_serializer(data=request.data)
        serializer.is_valid(raise_exception=True)

        game = Game.objects.get(id=serializer.data.get('game'))

        headers = self.get_success_headers(serializer.data)

        # Check if user doesn't already control a player in this game
        if game.players.filter(user__username=request.user.username):
            return Response(serializer.data, status=status.HTTP_403_FORBIDDEN, headers=headers)

        # else
        Player.objects.create(user=request.user, game=game)
        return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)`

Reply:
Je suis d'accord avec toi que devoir fournir son propre user_id dans la requête post pour créer un player est un peu étrange, mais dans l'état actuel on ne peux pas faire autrement. En effet, `PlayerViewSet` utilise le `PlayerSerializer`, qui a `user` dans ses `fields`, donc si j'ai bien compris, vu que user est un champs obligatoire, le `serializer.is_valid(raise_exception=True)` va renvoyer une erreur 400 si `serializer.data` ne possède pas de champs user.
Pour résoudre ce problème on pourrait:
- laisser tel quel, ce qui est ok mais un peu étrange et redondant
- supprimer la ligne `serializer.is_valid(raise_exception=True)`, ce qui me paraît une très mauvaise idée :)
- créer une autre vue `JoinGameViewSet` associé à un serializer `JoinGameSerializer` qui ne contient uniquement un champs game (et id), ce qui me paraît un peu mieux

Qu'en penses-tu ?

Reply:
J'ai une quatrième idée que je ne peux pas tester car je suis dans le train : il suffitde rendre le champ `user` read-only dans le serializer. Ça ne devrait pas poser problème puisqu'on crée un Player à la main et qu'on n'a aucune raison de permettre la modification du champ `user` après coup :p

Reply:
Ça semble fonctionner :)